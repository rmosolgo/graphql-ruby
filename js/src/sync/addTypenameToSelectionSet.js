/* eslint-disable max-len */
// This file/concept was borrowed from https://github.com/apollographql/apollo-client/blob/fec6457746b1cb63c759f128349e66499328ae43/src/queries/queryTransform.ts#L22-L56
/* eslint-enable max-len */

const TYPENAME_FIELD = {
  kind: 'Field',
  name: {
    kind: 'Name',
    value: '__typename',
  },
};

function addTypenameToSelectionSet(selectionSet, isRoot) {
  if (selectionSet.selections) {
    if (!isRoot) {
      const alreadyHasThisField = selectionSet.selections.some(selection => (
        selection.kind === 'Field' && selection.name.value === '__typename'
      ));

      if (!alreadyHasThisField) {
        selectionSet.selections.push(TYPENAME_FIELD);
      }
    }

    selectionSet.selections.forEach((selection) => {
      // Must not add __typename if we're inside an introspection query
      if (selection.kind === 'Field') {
        if (selection.name.value.lastIndexOf('__', 0) !== 0 && selection.selectionSet) {
          addTypenameToSelectionSet(selection.selectionSet, false);
        }
      } else if (selection.kind === 'InlineFragment') {
        if (selection.selectionSet) {
          addTypenameToSelectionSet(selection.selectionSet, false);
        }
      }
    });
  }
}

module.exports = addTypenameToSelectionSet;
