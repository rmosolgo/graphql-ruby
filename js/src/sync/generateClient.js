/**
 * Given a map of { name => alias } pairs,
 * generate a JavaScript module that that
 * takes names as input and returns the alias.
 * @param {String} clientName - the client ID that this map belongs to
 * @param {Object} nameToAlias - `name => alias` pairs
 * @return {String} generated JS code
*/
function generateMap(clientName, nameToAlias) {
  if (!clientName) {
    throw new Error("Client name is required to generate a persisted alias lookup map")
  }
  // Build up the map
  var keyValuePairs = "{"
  var operationName;
  for (operationName in nameToAlias) {
    var persistedAlias = nameToAlias[operationName]
    keyValuePairs += "\n  \"" + operationName + "\": \"" + persistedAlias + "\","
  }
  keyValuePairs += "\n}"

  // Insert the map, plus a function to fetch values from it
  var javaScriptCode = `/**
 * Generated by graphql-ruby-client
 *
*/

/**
 * Map local operation names to persisted keys on the server
 * @return {Object}
 * @private
*/
var _aliases = ${keyValuePairs}

/**
 * The client who synced these operations with the server
 * @return {String}
 * @private
*/
var _client = "${clientName}"

var OperationStoreClient = {
  /**
   * Build a string for \`params[:operationId]\`
   * @param {String} operationName
   * @return {String} stored operation ID
  */
  getOperationId: function(operationName) {
    return _client + "/" + OperationStoreClient.getPersistedQueryAlias(operationName)
  },

  /**
   * Fetch a persisted alias from a local operation name
   * @param {String} operationName
   * @return {String} persisted alias
  */
  getPersistedQueryAlias: function(operationName) {
    var persistedAlias = _aliases[operationName]
    if (!persistedAlias) {
      throw new Error("Failed to find persisted alias for operation name: " + operationName)
    } else {
      return persistedAlias
    }
  },

  /**
   * Satisfy the Apollo middleware API.
   * Replace the query with an operationId
  */
  apolloMiddleware: {
    applyMiddleware: function(options, next) {
      var req = options.request
      // Fetch the persisted alias for this operation
      req.operationId = OperationStoreClient.getOperationId(req.operationName)
      // Remove the now-unused query string
      delete req.query
      // Continue the request
      next()
    }
  }
}

module.exports = OperationStoreClient
`

  return javaScriptCode
}

module.exports = generateMap
