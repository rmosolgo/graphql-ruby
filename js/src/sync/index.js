const md5 = require('./md5');
const sendPayload = require('./sendPayload');
const prepareRelay = require('./prepareRelay');
const prepareIsolatedFiles = require('./prepareIsolatedFiles');
const prepareProject = require('./prepareProject');
const generateClient = require('./generateClient');
const printResponse = require('./printResponse');
const Logger = require('./logger');

const glob = require('glob');
const fs = require('fs');

/**
 * Find `.graphql` files in `path`,
 * then prepare them & send them to the configured endpoint.
 *
 * @param {Object} options
 * @param {String} options.path - A glob to recursively search for `.graphql` files (Default is `./`)
 * @param {String} options.secret - HMAC-SHA256 key which must match the server secret (default is no encryption)
 * @param {String} options.url - Target URL for sending prepared queries
 * @param {String} options.mode - If `"file"`, treat each file separately. If `"project"`, concatenate all files and
 * extract each operation. If `"relay"`, treat it as relay-compiler output
 * @param {Boolean} options.addTypename - Indicates if the "__typename" field are automatically added to your queries
 * @param {String} options.outfile - Where the generated code should be written
 * @param {String} options.client - the Client ID that these operations belong to
 * @param {Function} options.send - A function for sending the payload to the server, with the signature
 * `options.send(payload)`. (Default is an HTTP `POST` request)
 * @param {Function} options.hash - A custom hash function for query strings with the signature `options.hash(string)
 *  => digest` (Default is `md5(string) => digest`)
 * @return {void}
*/
function sync(options = {}) {
  const logger = new Logger(options.quiet);

  const { url } = options;
  if (!url) {
    throw new Error('URL must be provided for sync');
  }
  let graphqlGlob = options.path || './';
  const hashFunc = options.hash || md5;
  const encryptionKey = options.secret;
  if (encryptionKey) {
    logger.log('Authenticating with HMAC');
  }
  const sendFunc = options.send || sendPayload;
  const filesMode = options.mode || (graphqlGlob.indexOf('__generated__') > -1 ? 'relay' : 'project');

  let outfile;
  if (options.outfile) {
    /* eslint-disable prefer-destructuring */
    outfile = options.outfile;
    /* eslint-enable prefer-destructuring */
  } else if (fs.existsSync('src')) {
    outfile = 'src/OperationStoreClient.js';
  } else {
    outfile = 'OperationStoreClient.js';
  }

  const clientName = options.client;
  if (!clientName) {
    throw new Error('Client name must be provided for sync');
  }

  // Check for file ext already, add it if missing
  const containsFileExt = graphqlGlob.indexOf('.graphql') > -1 || graphqlGlob.indexOf('.gql') > -1;
  if (!containsFileExt) {
    graphqlGlob += '**/*.graphql*';
  }

  const payload = {
    operations: [],
  };

  const filenames = glob.sync(graphqlGlob, {});

  if (filesMode === 'relay') {
    payload.operations = prepareRelay(filenames);
  } else {
    if (filesMode === 'file') {
      payload.operations = prepareIsolatedFiles(filenames, options.addTypename);
    } else if (filesMode === 'project') {
      payload.operations = prepareProject(filenames, options.addTypename);
    } else {
      throw new Error(`Unexpected mode: ${filesMode}`);
    }
    // Update the operations with the hash of the body
    payload.operations.forEach((op) => {
      /* eslint-disable no-param-reassign */
      op.alias = hashFunc(op.body);
      /* eslint-enable no-param-reassign */
    });
  }

  if (payload.operations.length === 0) {
    logger.log(`No operations found in ${graphqlGlob}, not syncing anything`);
  } else {
    logger.log(`Syncing ${payload.operations.length} operations to ${logger.bright(url)}...`);

    const writeArtifacts = (response) => {
      const nameToAliasMap = {};
      const aliasToNameMap = {};
      payload.operations.forEach((op) => {
        nameToAliasMap[op.name] = op.alias;
        aliasToNameMap[op.alias] = op.name;
      });

      let responseData;

      if (response) {
        try {
          responseData = JSON.parse(response);
          printResponse(responseData, aliasToNameMap, logger);
          if (responseData.failed.length) {
            return false;
          }
        } catch (err) {
          logger.log('Failed to print sync result:', err);
        }
      }

      const generatedCode = generateClient(clientName, nameToAliasMap);
      logger.log(`Generating client module in ${logger.colorize('bright', outfile)}...`);
      fs.writeFileSync(outfile, generatedCode, 'utf8');
      logger.log(logger.green('âœ“ Done!'));

      return null;
    };

    const sendOpts = {
      url,
      client: clientName,
      secret: encryptionKey,
    };
    const maybePromise = sendFunc(payload, sendOpts);

    if (maybePromise instanceof Promise) {
      return maybePromise.then(writeArtifacts).catch((err) => {
        logger.error(logger.colorize('red', 'Sync failed:'));
        logger.error(err);
        return false;
      });
    }
    return writeArtifacts();
  }

  return null;
}

module.exports = sync;
