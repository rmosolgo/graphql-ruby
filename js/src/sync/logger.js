/* eslint-disable */
function Logger(isQuiet) {
  this.isQuiet = isQuiet;
}

Logger.prototype.log = function () {
  this.isQuiet ? null : console.log(...arguments);
};

Logger.prototype.error = function () {
  this.isQuiet ? null : console.error(...arguments);
};

const colors = {
  yellow: '\x1b[33m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
};

Logger.prototype.colorize = function (color, text) {
  const prefix = colors[color];
  if (!prefix) {
    throw new Error(`No color named: ${color}`);
  }
  return prefix + text + colors.reset;
};

function addColorizeShortcut(color) {
  Logger.prototype[color] = function (text) {
    return this.colorize(color, text);
  };
}
let colorName;
for (colorName in colors) {
  addColorizeShortcut(colorName);
}

module.exports = Logger;
