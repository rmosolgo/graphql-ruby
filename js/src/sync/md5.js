var crypto = require("crypto")

// Return the hex-encoded md5 hash of `inputString`
function md5(inputString) {
  return crypto.createHash("md5")
    .update(inputString)
    .digest("hex")
}

module.exports = md5
