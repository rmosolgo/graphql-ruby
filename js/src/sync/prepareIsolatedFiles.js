const fs = require('fs');
const graphql = require('graphql');
const addTypenameToSelectionSet = require('./addTypenameToSelectionSet');

/**
 * Read a bunch of GraphQL files and treat them as islands.
 * Don't join any fragments from other files.
 * Don't make assertions about name uniqueness.
 *
 */
function prepareIsolatedFiles(filenames, addTypename) {
  return filenames.map((filename) => {
    const fileOperationBody = fs.readFileSync(filename, 'utf8');
    let fileOperationName = null;

    const ast = graphql.parse(fileOperationBody);
    const visitor = {
      OperationDefinition: {
        enter(node) {
          if (fileOperationName) {
            /* eslint-disable max-len */
            throw new Error(`Found multiple operations in ${filename}: ${fileOperationName}, ${node.name}. Files must contain only one operation`);
            /* eslint-enable max-len */
          } else {
            fileOperationName = node.name.value;
          }
        },
        leave(node) {
          if (addTypename) { addTypenameToSelectionSet(node.selectionSet, true); }
        },
      },
      FragmentDefinition: {
        leave(node) {
          if (addTypename) { addTypenameToSelectionSet(node.selectionSet, true); }
        },
      },
    };
    graphql.visit(ast, visitor);

    return {
      // populate alias later, when hashFunc is available
      alias: null,
      name: fileOperationName,
      body: graphql.print(ast),
    };
  });
}

module.exports = prepareIsolatedFiles;
