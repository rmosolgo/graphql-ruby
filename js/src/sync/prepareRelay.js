const path = require('path');
const fs = require('fs');

/**
 * Read relay-compiler output
 * and extract info for persisting them & writing a map:
 *
 * - alias: get the relayHash from the header
 * - name: get the name from the JavaScript object
 * - body: get the text from the JavaScript object
 *
 * @param {Array} filenames -  Filenames to read
 * @return {Array} List of operations to persist & write to a map
 */
function prepareRelay(filenames) {
  const currentDirectory = process.cwd();
  const operations = [];
  filenames.forEach((filename) => {
    // Require the file to get values from the JavaScript code
    const absoluteFilename = path.resolve(currentDirectory, filename);
    /* eslint-disable global-require, import/no-dynamic-require */
    const operation = require(absoluteFilename);
    /* eslint-enable global-require, import/no-dynamic-require */
    const operationBody = operation.text;
    const operationName = operation.name;

    // Search the file for the relayHash
    const textContent = fs.readFileSync(filename, 'utf8');
    let operationAlias = textContent.match(/@relayHash ([a-z0-9]+)/);
    // Only operations get `relayHash`, so
    // skip over generated fragments
    if (operationAlias) {
      /* eslint-disable prefer-destructuring */
      operationAlias = operationAlias[1];
      /* eslint-enable prefer-destructuring */
      operations.push({
        alias: operationAlias,
        name: operationName,
        body: operationBody,
      });
    }
  });

  return operations;
}

module.exports = prepareRelay;
