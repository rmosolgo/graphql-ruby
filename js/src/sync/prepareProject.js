import fs from 'fs';
import { parse } from 'graphql';
import { print, visit } from 'graphql/language';
import addTypenameToSelectionSet from './addTypenameToSelectionSet';

// Return a new AST which contains only `definitionNames`
function extractDefinitions(ast, definitionNames) {
  const removeDefinitionNode = (node) => {
    if (definitionNames.indexOf(node.name.value) === -1) {
      return null;
    }
    return node;
  };
  const visitor = {
    OperationDefinition: removeDefinitionNode,
    FragmentDefinition: removeDefinitionNode,
  };

  const newAST = visit(ast, visitor);
  return newAST;
}

/**
 * Take a whole bunch of GraphQL in one big string
 * and validate it, especially:
 *
 * - operation names are unique
 * - fragment names are unique
 *
 * Then, split each operation into a free-standing document,
 * so it has all the fragments it needs.
 */

function prepareProject(filenames, addTypename) {
  let allGraphQL = '';
  filenames.forEach((filename) => {
    allGraphQL += fs.readFileSync(filename);
  });

  const ast = parse(allGraphQL);

  // This will contain { name: [name, name] } pairs
  const definitionDependencyNames = {};
  const allOperationNames = [];
  let currentDependencyNames = null;

  // When entering a fragment or operation,
  // start recording its dependencies
  const enterDefinition = (node) => {
    const definitionName = node.name.value;
    if (definitionDependencyNames[definitionName]) {
      /* eslint-disable max-len */
      throw new Error(`Found duplicate definition name: ${definitionName}, fragment & operation names must be unique to sync`);
      /* eslint-enable max-len */
    } else {
      /* eslint-disable no-multi-assign */
      currentDependencyNames = definitionDependencyNames[definitionName] = [];
      /* eslint-enable no-multi-assign */
    }
  };
  // When leaving, clean up, just in case
  const leaveDefinition = () => {
    currentDependencyNames = null;
  };

  const visitor = {
    OperationDefinition: {
      enter(node) {
        enterDefinition(node);
        allOperationNames.push(node.name.value);
      },
      leave(node) {
        if (addTypename) { addTypenameToSelectionSet(node.selectionSet, true); }
        return leaveDefinition;
      },
    },
    FragmentDefinition: {
      enter: enterDefinition,
      leave(node) {
        if (addTypename) { addTypenameToSelectionSet(node.selectionSet, true); }
        return leaveDefinition;
      },
    },
    // When entering a fragment spread, register it as a
    // dependency of its context
    FragmentSpread: {
      enter(node) {
        currentDependencyNames.push(node.name.value);
      },
    },
  };

  // Find the dependencies, build the accumulator
  visit(ast, visitor);

  // For each operation, build a separate document of that operation and its deps
  // then print the new document to a string
  const operations = allOperationNames.map((operationName) => {
    const visitedDepNames = [];
    const depNamesToVisit = [operationName];

    let depName;
    while (depNamesToVisit.length > 0) {
      depName = depNamesToVisit.shift();
      visitedDepNames.push(depName);
      definitionDependencyNames[depName].forEach((nextDepName) => {
        if (visitedDepNames.indexOf(nextDepName) === -1) {
          depNamesToVisit.push(nextDepName);
        }
      });
    }
    const newAST = extractDefinitions(ast, visitedDepNames);

    return {
      name: operationName,
      body: print(newAST),
      alias: null, // will be filled in later, when hashFunc is available
    };
  });

  return operations;
}


module.exports = prepareProject;
