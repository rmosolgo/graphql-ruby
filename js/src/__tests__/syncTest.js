import sync from '../sync';
import fs from 'fs';
import nock from 'nock';
import Logger from '../sync/logger';
var logger = new Logger

describe("sync operations", () => {
  beforeEach(() => {
    global.console.error = jest.fn()
    global.console.log = jest.fn()
  })

  afterEach(() => {
    jest.clearAllMocks();
  })

  describe("custom HTTP options", () => {
    it("uses the provided `send` option & provided URL", () => {
      var url = null
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        send: (sendPayload, options) => {
          url = options.url
        },
      }
      sync(options)

      expect(url).toEqual("bogus")
    })
  })

  describe("custom file processing options", () => {
    it("Adds .graphql to the glob if needed", () => {
      var payload = null
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        send: (sendPayload, opts) => { payload = sendPayload },
      }
      sync(options)

      expect(payload.operations).toMatchSnapshot()

      options.glob += "**/*.graphql"
      sync(options)

      // Get the same result, even when the glob already has a file extension
      expect(payload.operations).toMatchSnapshot()
    })

    it("Uses a custom hash function if provided", () => {
      var payload = null
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        hash: (graphQLBody) => {
          // This is a bad hack to get the operation name
          return graphQLBody.match(/query ([A-Za-z]+) \{/)[1].toUpperCase()
        },
        send: (sendPayload, opts) => { payload = sendPayload },
      }
      sync(options)

      expect(payload.operations).toMatchSnapshot()
    })
  })

  describe("Relay support", () => {
    it("Uses Relay output", () => {
      var payload = null
      var options = {
        client: "test-1",
        quiet: true,
        path: "./__generated__",
        url: "bogus",
        send: (sendPayload, opts) => { payload = sendPayload },
      }
      sync(options)

      expect(payload.operations).toMatchSnapshot()
    })
  })

  describe("Input files", () => {
    it("Merges fragments and operations across files", () => {
      var payload = null
      var options = {
        client: "test-1",
        quiet: true,
        path: "./src/__tests__/project/",
        url: "bogus",
        // mode: "project" is the default
        send: (sendPayload, opts) => { payload = sendPayload },
      }
      sync(options)

      expect(payload.operations).toMatchSnapshot()
    })

    it("Uses mode: file to process each file separately", () => {
      var payload = null
      var options = {
        client: "test-1",
        quiet: true,
        path: "./src/__tests__/project/",
        url: "bogus",
        mode: "file",
        send: (sendPayload, opts) => { payload = sendPayload },
      }
      sync(options)

      expect(payload.operations).toMatchSnapshot()
    })
  })

  describe("Sync output", () => {
    it("Generates a usable artifact for middleware", () => {
      var options = {
        client: "test-1",
        path: "./src/__tests__/project/",
        url: "bogus",
        quiet: true,
        send: (sendPayload, opts) => { },
      }
      sync(options)

      var generatedCode = fs.readFileSync("./src/OperationStoreClient.js", "utf8")
      expect(generatedCode).toMatch('"GetStuff": "5f0da489cf508a7c65ff5fa144e50545"')
      expect(generatedCode).toMatch('module.exports = OperationStoreClient')
      expect(generatedCode).toMatch('var _client = "test-1"')
      fs.unlinkSync("./src/OperationStoreClient.js")
    })

    it("Takes an outfile option", () => {
      var options = {
        client: "test-2",
        path: "./src/__tests__/project/",
        url: "bogus",
        quiet: true,
        outfile: "__crazy_outfile.js",
        send: (sendPayload, opts) => { },
      }
      sync(options)

      var generatedCode = fs.readFileSync("./__crazy_outfile.js", "utf8")
      expect(generatedCode).toMatch('"GetStuff": "5f0da489cf508a7c65ff5fa144e50545"')
      expect(generatedCode).toMatch('module.exports = OperationStoreClient')
      expect(generatedCode).toMatch('var _client = "test-2"')
      fs.unlinkSync("./__crazy_outfile.js")
    })
  })

  describe("Logging", () => {
    it("Logs progress", () => {
      var spy = console.log

      var options = {
        client: "test-1",
        path: "./src/__tests__/project/",
        url: "bogus",
        send: (sendPayload, opts) => { },
      }
      sync(options)

      expect(spy.mock.calls).toMatchSnapshot()
    })

    it("Can be quieted with quiet: true", () => {
      var spy = console.log

      var options = {
        client: "test-1",
        path: "./src/__tests__/project/",
        url: "bogus",
        quiet: true,
        send: (sendPayload, opts) => { },
      }
      sync(options)

      expect(spy.mock.calls).toMatchSnapshot()
    })
  })

  describe("Printing the result", () => {
    function buildMockRespondingWith(status, data) {
      return nock("http://example.com").post("/stored_operations/sync").reply(status, data)
    }

    it("prints failure", () => {
      var spyConsoleLog = console.log
      var spyConsoleError = console.error

      var mock = buildMockRespondingWith(422, {
        errors: { "5f0da489cf508a7c65ff5fa144e50545": ["something"] },
        failed: ["5f0da489cf508a7c65ff5fa144e50545"],
        added: ["defg"],
        not_modified: [],
      })

      var options = {
        client: "test-1",
        path: "./src/__tests__/project/",
        url: "http://example.com/stored_operations/sync",
        quiet: false,
      }

      var syncPromise = sync(options)

      return syncPromise.then(() => {
        expect(spyConsoleLog.mock.calls).toMatchSnapshot()
        expect(spyConsoleError.mock.calls).toMatchSnapshot()
        jest.clearAllMocks();
      })
    })

    it("prints success", () => {
      var spyConsoleLog = console.log
      var spyConsoleError = console.error

      var mock = buildMockRespondingWith(422, {
        errors: {},
        failed: [],
        added: ["defg"],
        not_modified: ["xyz", "123"],
      })

      var options = {
        client: "test-1",
        path: "./src/__tests__/project/",
        url: "http://example.com/stored_operations/sync",
        quiet: false,
      }

      var syncPromise = sync(options)

      expect(spyConsoleLog.mock.calls).toMatchSnapshot()
      jest.clearAllMocks();

      return syncPromise.then(() => {
        expect(spyConsoleLog.mock.calls).toMatchSnapshot()
        expect(spyConsoleError.mock.calls).toMatchSnapshot()
        jest.clearAllMocks();
      })
    })
  })
})
