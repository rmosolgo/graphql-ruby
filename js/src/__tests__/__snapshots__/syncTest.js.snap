// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sync operations Input files Merges fragments and operations across files 1`] = `
Array [
  Object {
    "alias": "5f0da489cf508a7c65ff5fa144e50545",
    "body": "fragment Frag1 on Query {
  moreStuff
}

query GetStuff {
  ...Frag1
}
",
    "name": "GetStuff",
  },
  Object {
    "alias": "c944b08d15eb94cf93dd124b7d664b62",
    "body": "fragment Frag1 on Query {
  moreStuff
}

fragment Frag2 on Query {
  ...Frag3
}

fragment Frag3 on Query {
  evenMoreStuff
}

query GetStuff2 {
  stuff
  ...Frag1
  ...Frag2
}
",
    "name": "GetStuff2",
  },
  Object {
    "alias": "fc215a466a3ea309fe493e8f7cb206f2",
    "body": "fragment Frag2 on Query {
  ...Frag3
}

fragment Frag3 on Query {
  evenMoreStuff
}

fragment Frag4 on Query {
  evenMoreStuff {
    stuffInside
  }
}

query GetStuff3 {
  stuff {
    withStuffInside
  }
  ...Frag2
  ...Frag4
}
",
    "name": "GetStuff3",
  },
  Object {
    "alias": "919af8f80d8848a9b9b6176e457f8e45",
    "body": "query GetStuffIsolated {
  ...FragIsolated
  things {
    existHere
  }
}

fragment FragIsolated on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": "GetStuffIsolated",
  },
  Object {
    "alias": "d5dfe825034aeda06e2639935d8b107d",
    "body": "query GetStuffIsolated2 {
  things {
    existHere
  }
}
",
    "name": "GetStuffIsolated2",
  },
]
`;

exports[`sync operations Input files Uses mode: file to process each file separately 1`] = `
Array [
  Object {
    "alias": "26c44bfe42872860da112b6177355bfa",
    "body": "fragment Frag1 on Query {
  moreStuff
}
",
    "name": null,
  },
  Object {
    "alias": "7de9e7bf1d6ea1f527de07a25983086c",
    "body": "fragment Frag2 on Query {
  ...Frag3
}
",
    "name": null,
  },
  Object {
    "alias": "8bc9b9922a7fbb66f4ac6c58d5d5c357",
    "body": "fragment Frag3 on Query {
  evenMoreStuff
}
",
    "name": null,
  },
  Object {
    "alias": "bfc521a5ade5ff6f17bdf9352c86c850",
    "body": "fragment Frag4 on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": null,
  },
  Object {
    "alias": "0a6add7303775e2487f2c2235ecb1c80",
    "body": "query GetStuff {
  ...Frag1
}
",
    "name": "GetStuff",
  },
  Object {
    "alias": "2f26b770ded2a04279bc4bf824ca54ac",
    "body": "query GetStuff2 {
  stuff
  ...Frag1
  ...Frag2
}
",
    "name": "GetStuff2",
  },
  Object {
    "alias": "1b8da77aabef67ee54df7e5acfd75893",
    "body": "query GetStuff3 {
  stuff {
    withStuffInside
  }
  ...Frag2
  ...Frag4
}
",
    "name": "GetStuff3",
  },
  Object {
    "alias": "919af8f80d8848a9b9b6176e457f8e45",
    "body": "query GetStuffIsolated {
  ...FragIsolated
  things {
    existHere
  }
}

fragment FragIsolated on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": "GetStuffIsolated",
  },
  Object {
    "alias": "d5dfe825034aeda06e2639935d8b107d",
    "body": "query GetStuffIsolated2 {
  things {
    existHere
  }
}
",
    "name": "GetStuffIsolated2",
  },
]
`;

exports[`sync operations Logging Can be quieted with quiet: true 1`] = `Array []`;

exports[`sync operations Logging Logs progress 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mbogus[0m...",
  ],
  Array [
    "Generating client module in [1msrc/OperationStoreClient.js[0m...",
  ],
  Array [
    "[32mâœ“ Done![0m",
  ],
]
`;

exports[`sync operations Printing the result prints failure 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mhttp://example.com/stored_operations/sync[0m...",
  ],
  Array [
    "  [2m0 added[0m",
  ],
  Array [
    "  [2m0 not modified[0m",
  ],
  Array [
    "  [31m1 failed[0m",
  ],
]
`;

exports[`sync operations Printing the result prints failure 2`] = `
Array [
  Array [
    "Sync failed, errors:",
  ],
  Array [
    "  GetStuff:",
  ],
  Array [
    "    [31mâœ˜[0m something",
  ],
]
`;

exports[`sync operations Printing the result prints success 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mhttp://example.com/stored_operations/sync[0m...",
  ],
]
`;

exports[`sync operations Printing the result prints success 2`] = `
Array [
  Array [
    "  [32m1 added[0m",
  ],
  Array [
    "  [0m2 not modified[0m",
  ],
  Array [
    "  [2m0 failed[0m",
  ],
  Array [
    "Generating client module in [1msrc/OperationStoreClient.js[0m...",
  ],
  Array [
    "[32mâœ“ Done![0m",
  ],
]
`;

exports[`sync operations Printing the result prints success 3`] = `Array []`;

exports[`sync operations Relay support Uses Relay output 1`] = `
Array [
  Object {
    "alias": "353e010cb78d082b29cb63ee7e9027b3",
    "body": "query AppFeedQuery {
  feed(type: NEW, limit: 5) {
    ...Feed
  }
}

fragment Feed on Entry {
  repository {
    owner {
      login
    }
    name
  }
  ...FeedEntry
}

fragment FeedEntry on Entry {
  repository {
    owner {
      login
    }
    name
    stargazers_count
  }
  postedBy {
    login
  }
}
",
    "name": "AppFeedQuery",
  },
]
`;

exports[`sync operations custom file processing options Adds .graphql to the glob if needed 1`] = `
Array [
  Object {
    "alias": "f7f65309043352183e905e1396e51078",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;

exports[`sync operations custom file processing options Adds .graphql to the glob if needed 2`] = `
Array [
  Object {
    "alias": "f7f65309043352183e905e1396e51078",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;

exports[`sync operations custom file processing options Uses a custom hash function if provided 1`] = `
Array [
  Object {
    "alias": "GETSTUFF",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;
